File: parity/oracle.mjs
#!/usr/bin/env node
// parity/oracle.mjs — Node-side parity harness.
//
// Runs the same deterministic scenario as `crates/sim/src/parity.rs` and
// emits a JSON snapshot for `cargo run --bin parity-check` to compare.
//
// Usage:
//   node parity/oracle.mjs SEED ENTITIES TICKS HBEVERY > oracle.json

import { SimulationKernel } from '../src/core/SimulationKernel.js';

const [, , seedStr = '12345', entitiesStr = '8', ticksStr = '200', hbStr = '5'] = process.argv;
const seed = Number(seedStr);
const entities = Number(entitiesStr);
const ticks = Number(ticksStr);
const heartbeatEvery = Number(hbStr);

const k = new SimulationKernel(seed);

// Spawn entities at deterministic positions (matches parity.rs exactly).
for (let i = 0; i < entities; i++) {
  const template = {
    type: i % 2 === 0 ? 'marker' : 'person',
    position: { x: i * 1.5, y: i % 7, z: 0 },
    mass: i % 3 === 0 ? 1.0 : 0.0,
    isPerson: i % 2 === 1,
  };
  k.createEntity(template);
}

const rngFirst5 = [];
const rngLast5 = [];
const eventLogSample = [];
let rngCount = 0;

for (let i = 0; i < ticks; i++) {
  const r = k.random();
  rngCount++;
  if (rngFirst5.length < 5) rngFirst5.push(r);

  k.scheduleEvent({ type: 'parity_heartbeat', rng: r }, heartbeatEvery);
  k.tick();

  if (rngLast5.length === 5) rngLast5.shift();
  rngLast5.push(r);
}

const snapshot = {
  kind: 'parity',
  version: 1,
  seed,
  entities,
  ticks,
  heartbeatEvery,
  rngFirst5,
  rngLast5,
  rngCount,
  worldTime: {
    year: k.worldTime.year,
    day: k.worldTime.day,
    hour: k.worldTime.hour,
    minute: k.worldTime.minute,
  },
  turn: k.turn,
  alivePeople: k.alivePeople.size,
  conservation: k.conservationLedger,
  firstEntityIds: Array.from(k.entities.keys()).slice(0, 8),
  firstEntityPositions: Array.from(k.entities.values()).slice(0, 8).map(
    (e) => e.position ? { x: e.position.x, y: e.position.y, z: e.position.z ?? 0 } : null
  ),
  eventLogSample: k.eventLog.slice(0, 16).map((e) => ({
    id: e.id, turn: e.processedAt ?? e.turn ?? 0, type: e.type,
  })),
};

process.stdout.write(JSON.stringify(snapshot, null, 2) + '\n');

File: crates/sim/src/bin/oracle.rs
//! `oracle` — runs the parity scenario and emits JSON on stdout.
//! Used by the Node-side cross-check to compare against the Node oracle.
//!
//! Usage:
//!   cargo run --quiet --bin oracle -- [seed=12345] [entities=8] [ticks=200] [heartbeat_every=5] > out.json

use sim::parity::{run, ParityInput};
use std::env;
use std::process;

fn main() {
    let seed: u32 = env::args().nth(1).and_then(|s| s.parse().ok()).unwrap_or(12345);
    let entities: u32 = env::args().nth(2).and_then(|s| s.parse().ok()).unwrap_or(8);
    let ticks: u64 = env::args().nth(3).and_then(|s| s.parse().ok()).unwrap_or(200);
    let hb: u64 = env::args().nth(4).and_then(|s| s.parse().ok()).unwrap_or(5);

    let snap = run(ParityInput { seed, entities, ticks, heartbeat_every: hb });
    match serde_json::to_string_pretty(&snap) {
        Ok(s) => {
            println!("{s}");
            process::exit(0);
        }
        Err(e) => {
            eprintln!("oracle: failed to serialize: {e}");
            process::exit(1);
        }
    }
}
